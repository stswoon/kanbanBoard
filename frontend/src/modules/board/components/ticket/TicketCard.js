// @flow

import React, {Component} from "react";
import {Button, DatePicker, Input, Popconfirm} from "antd";
import "./TicketCard.less";
import moment from "moment";
import {strings} from "../../../shared/services/strings";
import {utils} from "../../../shared/services/utils";
import type {Ticket, UUID} from "../../BoardModels";

const debounce = require('lodash.debounce');
const {TextArea} = Input;

type Props = Ticket | {
    onChange: (ticket: Ticket) => void,
    onRemove: (ticketId: UUID) => void
};
type State = Ticket;

export class TicketCard extends Component<Props, State> {
    constructor(props) {
        super(props);
        this.state = {
            id: this.props.id,
            boardId: this.props.boardId,
            name: this.props.name,
            description: this.props.description,
            dueDate: this.props.dueDate,
            status: this.props.status,
            order: this.props.order
        };
    }

    render() {
        return (
            <div className="ticket">
                <div className="ticket__header"/>
                <div className="ticket__name">
                    <Input className="ticket__input" defaultValue={this.props.name || strings.ticket.noname}
                           name="name" onChange={this.handleInputChange}/>
                    <Popconfirm title={strings.ticket.areYouSure} okText={strings.ticket.remove}
                                cancelText={strings.ticket.cancel} onConfirm={this.remove}>
                        <Button type="danger" icon="delete"/>
                    </Popconfirm>
                </div>
                <div className="ticket__dueDate">
                    <span className="ticket__input-name">{strings.ticket.dueDate}</span>
                    <DatePicker className="ticket__input"
                                defaultValue={this.props.dueDate && moment(this.props.dueDate)}
                                onChange={this.handleDateChange} placeholder=""/>
                </div>
                <div className="ticket__description">
                    <TextArea className="ticket__input" rows={4} defaultValue={this.props.description}
                              name="description" onChange={this.handleInputChange}/>
                </div>
            </div>
        );
    }

    remove = () => {
        this.props.onRemove(this.props.id);
    };

    handleInputChange = (event) => {
        const value = event.target.value;
        const name = event.target.name;
        this.setState({[name]: value}, () => {
            this.handleChange();
        });
    };

    handleDateChange = (date, dateString) => {
        console.log(date, dateString);
        this.setState({dueDate: dateString}, () => {
            this.handleChange();
        });
    };

    debounceHandleChange;

    handleChange = () => {
        if (!this.debounceHandleChange) {
            let f = () => {
                this.props.onChange(utils.object.deepCopy(this.state));
            };
            f = f.bind(this);
            this.debounceHandleChange = debounce(f, 500);
        }
        this.debounceHandleChange();
    }
}