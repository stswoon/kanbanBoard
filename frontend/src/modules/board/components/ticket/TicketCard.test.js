import React from 'react';
import {shallow} from "enzyme";
import {shallowToJson} from 'enzyme-to-json';
import {TicketCard} from "./TicketCard";

const ticket = {
    "id": "d5e3adf4-d88e-426f-a734-aabe3f82eb99",
    "name": "Requirements",
    "description": "Lorem ipsum",
    "dueDate": "2019-03-07T00:00:00.000+0000",
    "status": "DONE"
};

it('renders without crashing', () => {
    const ar = shallow(<TicketCard {...ticket}/>);
    expect(shallowToJson(ar)).toMatchSnapshot();
});

