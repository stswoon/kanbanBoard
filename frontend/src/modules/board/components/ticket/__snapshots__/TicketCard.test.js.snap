// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="ticket"
>
  <div
    className="ticket__header"
  />
  <div
    className="ticket__name"
  >
    <Input
      className="ticket__input"
      defaultValue="Requirements"
      disabled={false}
      name="name"
      onChange={[Function]}
      type="text"
    />
    <Popconfirm
      cancelText="Cancel"
      icon={
        <Icon
          theme="filled"
          type="exclamation-circle"
        />
      }
      okText="Remove"
      okType="primary"
      onConfirm={[Function]}
      placement="top"
      title="Are you sure to remove ticket?"
      transitionName="zoom-big"
      trigger="click"
    >
      <Button
        block={false}
        ghost={false}
        icon="delete"
        loading={false}
        type="danger"
      />
    </Popconfirm>
  </div>
  <div
    className="ticket__dueDate"
  >
    <span
      className="ticket__input-name"
    >
      Due date:
    </span>
    <PickerWrapper
      className="ticket__input"
      defaultValue={"2019-03-07T00:00:00.000Z"}
      locale={Object {}}
      onChange={[Function]}
      onOk={[Function]}
      onOpenChange={[Function]}
      placeholder=""
      popupStyle={Object {}}
      transitionName="slide-up"
    />
  </div>
  <div
    className="ticket__description"
  >
    <TextArea
      className="ticket__input"
      defaultValue="Lorem ipsum"
      name="description"
      onChange={[Function]}
      rows={4}
    />
  </div>
</div>
`;
